Framework/Source/Graphics/Program/ParameterBlock.h

```diff
@@ -0,0 +1,272 @@
+
+        struct RootData
+        {
+            RootData() = default;
+            RootData(uint32_t root, uint32_t range) : rootIndex(root), rangeIndex(range) {}
+            uint32_t rootIndex = uint32_t(-1);
+            uint32_t rangeIndex = uint32_t(-1);
+        };
+
+        template<typename ViewType>
+        struct ResourceData
+        {
+            ResourceData(const RootData& data) : rootData(data) {}
+            typename ViewType::SharedPtr pView = nullptr;
+            Resource::SharedPtr pResource = nullptr;
+            RootData rootData;
+        };
+
+        template<>
+        struct ResourceData<Sampler>
+        {
+            ResourceData(const RootData& data) : rootData(data) {}
+            Sampler::SharedPtr pSampler;
+            RootData rootData;
+        };
+
+
+        struct RootSet
+        {
+            mutable std::shared_ptr<DescriptorSet> pDescSet;
+            mutable bool dirty = false;
+        };
+
+        union BindLocation
+        {
+            BindLocation() = default;
+            BindLocation(uint32_t space, uint32_t index) : baseRegIndex(index), regSpace(space) {}
+            struct
+            {
+                uint32_t baseRegIndex;
+                uint32_t regSpace;
+            };
+            uint64_t u64 = -1;
+            std::size_t operator()(BindLocation b) const { return std::hash<uint64_t>{}(u64); }
+            bool operator==(const BindLocation& other) const { return u64 == other.u64; }
+        };
+
+        template<typename T>
+        using ResourceMap = std::unordered_map<BindLocation, std::vector<ResourceData<T>>, BindLocation>;
+        using SamplerMap = std::unordered_map<BindLocation, std::vector<Sampler::SharedConstPtr>, BindLocation>;
+        using RootSetVec = std::vector<RootSet>;
+
+        const ResourceMap<ConstantBuffer>& getAssignedCbs() const { return mAssignedCbs; }
+        const ResourceMap<ShaderResourceView>& getAssignedSrvs() const { return mAssignedSrvs; }
+        const ResourceMap<UnorderedAccessView>& getAssignedUavs() const { return mAssignedUavs; }
+        const ResourceMap<Sampler>& getAssignedSamplers() const { return mAssignedSamplers; }
+        const RootSetVec getRootSets() const { return mRootSets; }
+
+        // Delete some functions. If they are not deleted, the compiler will try to convert the uints to string, resulting in runtime error
+        Sampler::SharedPtr getSampler(uint32_t) const = delete;
+        bool setSampler(uint32_t, const Sampler::SharedPtr&) = delete;
+        bool setConstantBuffer(uint32_t, const ConstantBuffer::SharedPtr&) = delete;
+        ConstantBuffer::SharedPtr getConstantBuffer(uint32_t) const = delete;
+    private:
+        ParameterBlock(const ParameterBlockReflection::SharedConstPtr& pReflection, const RootSignature* pRootSig, bool createBuffers);
+        ParameterBlockReflection::SharedConstPtr mpReflector;
+
+        ResourceMap<ConstantBuffer> mAssignedCbs;        // HLSL 'b' registers
```

Comment:
I know we've gone in circles on this before, so I won't belabor it, but it seems to me that a `ParameterBlock` should be able to store all of its bound resources in nice contiguous arrays (or `vector`s) so that operations that are based on the "local" indexing inside the parameter block can be as efficient as possible.

That is, if I call `setSampler(someIndex, someSampler)` it should ideally translate almost directly into `mAssignedSamplers[someIndex] = someSampler`, where `someIndex` is always between zero and `mpReflector->getSamplerCount()`.

Ideally those contiguous arrays are also set up so that when populating a descriptor set/table with those values, you can just pass `&mAssignedSamplers[0]` off to the graphics API (I know it isn't *quite* that simple).

I understand why you can't easily do it this way: you want to preserve an API where the user can just pass in the value they put in a `register` modifier as the index and expect that to Just Work, but at some point that forces extra indirections which means you aren't going to be as fast as possible.

Reply:
> I understand why you can't easily do it this way: you want to preserve an API where the user can just pass in the value they put in a register modifier as the index and expect that to Just Work, but at some point that forces extra indirections which means you aren't going to be as fast as possible.

Yup, people want to use the space and index they put in the HLSL. Not SOL, but simpler to use.